devise を使ってユーザー認証を実装

sjabcdefin opened the pull request:
「devise を使ってユーザー認証を実装する」の課題を実施しました。
ご確認お願いいたします。

sjabcdefin committed:
Install devise gem

deviseをインストール

sjabcdefin committed:
Create User model of devise

deviseのユーザモデルを作成

sjabcdefin committed:
Create devise:controllers users

deviseコントローラの作成

sjabcdefin committed:
Implement functions of Sign up, Sign in and Log out

サインアップ、サインイン、ログアウトの機能を実装

sjabcdefin committed:
Create users controller and Add resources :users to routes.rb

- usersコントローラの作成
- routes.rbに、resources :usersを追加

sjabcdefin committed:
Implement functions of user list, details and editing

ユーザ一覧、詳細、編集機能を実装

sjabcdefin committed:
Implement changing password

パスワード変更機能を実装

sjabcdefin committed:
Implement the function of paging in User list page

ユーザ一覧画面にページング処理を実装

sjabcdefin committed:
Remove unnecessary files for customization

カスタマイズ不要なファイルを削除

sjabcdefin committed:
Implemented Japanese localization

日本語化を実施

sjabcdefin committed:
Implement tests

テストを実装

sjabcdefin committed:
Add root path to config/routes.rb

config/routes.rbにroot pathの定義を追加

sjabcdefin committed:
Modify the rubocop and erblint indicated item

rubocopとerblintの指摘事項を修正

sjabcdefin committed:
Delete unnecessary files

不要なファイルを削除

sjabcdefin committed:
update: Correct update method name of before_action

before_actionのupdateメソッド名称を修正

yuuu commented on app/controllers/users/passwords_controller.rb:
```diff
@@ -0,0 +1,14 @@
+class Users::PasswordsController < Devise::PasswordsController
+  protected
+
+  def after_resetting_password_path_for(*)
```
このあたりの指定は必要でしたでしょうか？記憶が間違っているかもしれませんが、このメソッドを定義せずともパスワードをリセットした後のリダイレクト先はログインページになっていたような気がしています。

sjabcdefin replied:
ご指摘の通り、このメソッドを定義せずとも、パスワードリセット後・パスワードメール送付後のリダイレクト先はログインページになっていました。
本ファイルは、リダイレクト先のみの指定を記載していたため、ファイル自体を削除しました。

修正コミット：https://github.com/sjabcdefin/fjord-books_app-2023/pull/3/commits/a2342a4d0fe530e0a0ce92dc81dd9129956d18f1

yuuu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
ここはdeviseから送信されるメールのfromに該当する部分なので `no-reply@example.com` のようなメールアドレスを設定しておきましょう。

sjabcdefin replied:
```no-reply@example.com```を設定いたしました。

修正コミット：https://github.com/sjabcdefin/fjord-books_app-2023/pull/3/commits/580a68ee35517d8a1a001e388a131f27acb70a3d

yuuu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
ここはdeviseから送信されるメールのfromに該当する部分なので `no-reply@example.com` のようなメールアドレスを設定しておきましょう。

yuuu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+
+  # ==> Configuration for :validatable
+  # Range for password length.
+  config.password_length = 6..128
```
昨今のセキュリティ情勢を踏まえると6文字のパスワードは心もとないかもしれません。最小8文字としてはどうでしょうか？

sjabcdefin replied:
最小8文字に修正いたしました。

修正コミット：https://github.com/sjabcdefin/fjord-books_app-2023/pull/3/commits/24fe04eecdb26f977f16a1a9acbef36cc5729aed, https://github.com/sjabcdefin/fjord-books_app-2023/pull/3/commits/4978a68f64bfc1a21193e5dc4e2f0bd5e138f758

yuuu commented on app/controllers/books_controller.rb:
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 class BooksController < ApplicationController
+  before_action :authenticate_user!
```
今回のアプリケーションは基本ログインが必須になるので `ApplicationController` にてこの `before_action` を記載すると、今後コントローラが増えたときの呼び出し忘れを防ぐことができます。

sjabcdefin replied:
```ApplicationController```に```before_action :authenticate_user!```を記載するように修正いたしました。

修正コミット：https://github.com/sjabcdefin/fjord-books_app-2023/pull/3/commits/846fff5347bf3f43250a266de0710820c4ad5cda

yuuu commented on config/routes.rb:
```diff
@@ -1,7 +1,20 @@
 
   # Defines the root path route ("/")
   # root "articles#index"
+  root to: redirect('/users/sign_in')
```
ここですが `root` のPathを `books#index` とするのはどうでしょうか？

sjabcdefin replied:
```root```Pathを```books#index```に修正いたしました。

修正コミット：https://github.com/sjabcdefin/fjord-books_app-2023/pull/3/commits/ee9296589a180c7317067cb8c1e94203514d9eb0, https://github.com/sjabcdefin/fjord-books_app-2023/pull/3/commits/137bf6dcd8cf794b2fa0a541c67de22f104d64d7

yuuu commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,9 @@
+
+FactoryBot.define do
+  factory :user do
+    email { 'test@example.com' }
```
この書き方だとテスト中でユーザーを複数作成できないので `sequence` を使っておくと良いですね。

sjabcdefin replied:
```sequence```を使用するように修正いたしました。

修正コミット：https://github.com/sjabcdefin/fjord-books_app-2023/pull/3/commits/ec8ae9d6abb2c11feac2d3bd6d459f90573cf0dd

sjabcdefin committed:
Delete: delete passwords_controller.rb

passwords_controller.rbを削除

sjabcdefin committed:
Update: Modify mail_sender in devise.rb

メール送信元のメールアドレスを変更

sjabcdefin committed:
Update: Modify password_length

パスワードの最低文字数を6から8に変更

sjabcdefin committed:
Update: Move authenticate_user! from BooksController to ApplicationController

authenticate_user!をBooksControllerからApplicationControllerに移動

sjabcdefin committed:
Update: Modify root path in routes.rb and Delete redirect path ('books_path') after sign-up/sign-in

rootパスを'books#index'に変更
サインアップ・サインイン後のリダイレクトパス('books_path')を削除

sjabcdefin committed:
Update: Modify spec file due to changing password length

パスワード文字数変更に伴うspecファイル修正

sjabcdefin committed:
Update: Modify path from 'books_path' to 'root_path' in spec files

specファイルの'books_path'を'root_path'に修正

sjabcdefin committed:
Update: Use sequence to create multiple users

複数ユーザを作成するためにsequenceを使用

Update: Use sequence to create multiple users(Correct missing corrections)

複数ユーザを作成するためにsequenceを使用(修正漏れを対応)
